src/Mvc/src/Microsoft.AspNetCore.Mvc.ViewFeatures/HtmlHelperOfT.cs

```diff
@@ -361,41 +349,72 @@ public string NameFor<TResult>(Expression<Func<TModel, TResult>> expression)
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            var modelExplorer = GetModelExplorer(expression);
+            var modelExpression = GetModelExpression(expression);
             return GenerateTextBox(
-                modelExplorer,
-                GetExpressionName(expression),
-                modelExplorer.Model,
+                modelExpression.ModelExplorer,
+                modelExpression.Name,
+                modelExpression.Model,
                 format,
                 htmlAttributes);
         }
 
-        protected string GetExpressionName<TResult>(Expression<Func<TModel, TResult>> expression)
+        /// <summary>
+        /// Gets the <see cref="ModelExpression"/> for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The <see cref="ModelExpression"/>.</returns>
+        public ModelExpression GetModelExpression<TResult>(Expression<Func<TModel, TResult>> expression)
         {
             if (expression == null)
             {
                 throw new ArgumentNullException(nameof(expression));
             }
 
-            return ExpressionHelper.GetExpressionText(expression, _expressionTextCache);
+            return _modelExpressionProvider.CreateModelExpression(ViewData, expression);
         }
 
-        protected ModelExplorer GetModelExplorer<TResult>(Expression<Func<TModel, TResult>> expression)
+        /// <summary>
+        /// Gets the <see cref="ModelExpression"/> for <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TResult">The type of the result.</typeparam>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The <see cref="ModelExpression"/>.</returns>
+        public ModelExpression GetModelExpression<TResult>(string expression)
```

Comment:
Again, this one-liner isn't worth making `public`. I wouldn't recommend any code external to this project use this method.